src/main/java/com/example/petbuddybackend/controller/ChatWebSocketController.java

```diff
@@ -38,23 +49,52 @@ public void sendChatMessage(
             @Headers Map<String, Object> headers,
             Principal principal
     ) {
-        String username = principal.getName();
+        String principalUsername = principal.getName();
         Role acceptRole = HeaderUtils.getHeaderSingleValue(headers, ROLE_HEADER_NAME, Role.class);
-        ChatMessageDTO messageDTO = chatService.createMessage(chatId, username, message, acceptRole);
+        ChatMessageDTO messageDTO = chatService.createMessage(chatId, principalUsername, message, acceptRole);
+        MessageCallback callback = chatService.createCallbackMessageSeen(chatId, principalUsername);
 
-        chatSessionService.patchMetadata(chatId, username, headers);
-        chatSessionService.sendMessages(chatId, messageDTO);
+        chatSessionService.patchMetadata(chatId, principalUsername, headers);
+        chatSessionService.sendNotifications(chatId, new ChatNotificationMessage(messageDTO), callback);
     }
 
     @EventListener
     public void handleSubscription(SessionSubscribeEvent event) {
-        StompHeaderAccessor headerAccessor = StompHeaderAccessor.wrap(event.getMessage());
-        chatSessionService.subscribeIfAbsent(headerAccessor);
+        StompHeaderAccessor accessor = StompHeaderAccessor.wrap(event.getMessage());
+        String timeZone = HeaderUtils.getHeaderSingleValue(accessor, TIMEZONE_HEADER_NAME, String.class);
+        String username = HeaderUtils.getUser(accessor);
+        Long chatId = HeaderUtils.getLongFromDestination(accessor, CHAT_ID_INDEX_IN_TOPIC_URL);
+
+        chatService.updateLastMessageSeen(chatId, username);
+        chatSessionService.subscribe(chatId, username, timeZone);
+        chatSessionService.sendNotifications(chatId, new ChatNotificationJoined(chatId, username));
     }
 
     @EventListener
-    public void handleUnsubscription(SessionUnsubscribeEvent event) {
-        StompHeaderAccessor headerAccessor = StompHeaderAccessor.wrap(event.getMessage());
-        chatSessionService.unsubscribeIfPresent(headerAccessor);
+    public void handleUnsubscribe(SessionUnsubscribeEvent event) {
+        ChatSessionContext context = chatSessionService.getContext();
```

Comment:
a to zwroci context akurat wczesniej zalogowanego uzytkownika? Bo ich moglo byc kilku. I w ChatSessionService mamy ChatSessionManagera, ktory trzyma kilka pokoi i uzytkownikow, a context mamy jeden chyba i w momencie nowej subskrybcji on sie zmienia. Ale jako ze jestesmy serwisem to kilku na raz moze subskrybowac. Chyba ze czegos nie rozumiem z contextem

Reply:
Ogólnie zacznę od tego, że ten context to powstał w związku z eventem na disconnect. Tutaj nie dostaniemy żadnych informacji od frontu, bo to jest case gdzie użytkownik np. traci połączenie internetowe. A przy disconnect'ie potrzebowałem chatId żeby usunąć informacje o użytkowniku.

Ten ChatSessionContext to bean ze scope na sesję websocketa (możesz sprawdzić klasę ChatSessionContext), więc dla każdej sesji jest osobny bean. Z domyślnym scopem, czyli Singleton to byłby faktycznie problem ale tutaj dla dwóch różnych użytkowników można różne dane przechowywać

Reply:
aa, czyli dla kazdego polaczenia jest osobny objekt Contextu, to git